keps/sig-network/3685-endpointslice-reconciler-to-staging/README.md

```diff
@@ -0,0 +1,154 @@
+code to `staging/src/k8s.io/endpointslice` and then import it as Go module.
+These changes include making private methods public and updating the import
+paths.
+
+## Motivation
+
+Moving the EndpointSlice reconciler to staging has several benefits:
+
+1. The reconciler logic can be reused by custom EndpointSlice controllers. The
+   fanout behavior is difficult to implement correctly and the current
+   reconciler code has gone through several iterations to reach its currently
+   level of robustness. There are cases where reconciler code has either been
+   forked or rewritten, both of which are problematic from a supportability
+   perspective.
+2. It reduces the burden of fully migrating from Endpoints to EndpointSlices.
+   It’s relatively easy to write a custom Endpoints controller, but less so with
+   EndpointSlices. So, the migration path is unclear for pre-existing custom
+   Endpoints controllers. Having a library that handles the more difficult
+   aspects of managing EndpointSlices would mitigate some of the risk in
+   migrating these controllers.
+3. It would help to realize one of the design goals of EndpointSlices: that
+   other controllers would exist to manage EndpointSlices, which is why the
+   `endpointslice.kubernetes.io/managed-by` annotation was added.
+
+### Goals
+
+- Expose a simple, and highly maintained, API for EndpointSlice reconciling.
+- Allow for the EndpointSlice reconciling code to easily be imported into other
+  projects.
+
+### Non-Goals
+
+- A generic solution for reconciling fan-out style resources.
+
+## Proposal
+
+1. Add the staging repository (`staging/src/k8s.io/endpointslice`) to
+   `kubernetes/kubernetes` by following the steps in the [staging directory
+   README][staging-readme].
+2. Move the reconciler submodule and it’s [dependencies](###dependencies) to the
+   staging directory and update imports in `pkg/controller/endpointslice`. This
+   should be a single atomic change.
+    - Some methods need to be made public to allow for reconciler code to be
+      used as a library (at least `NewReconciler` and `Reconcile`).
+3. Allow for the value of the `endpointslice.kubernetes.io/managed-by` label to
+   be configurable by adding `managedBy` as an argument to `NewReconciler`.
+4. Split computing the diff (EndpointSlices to create, update, and delete) and
+   applying the diff into separate methods. This allows for controllers to
+   easily add metadata to the managed EndpointSlices (e.g. adding an
+   annotation).
+
+[staging-readme]: https://github.com/kubernetes/kubernetes/tree/master/staging "External Repository Staging Area"
+
+### Dependencies
+
+The following dependencies would also have to be moved to the staging repository
+(assume `staging_root = staging/src/k8s.io/endpointslice`):
+- `pkg/controller/endpointslice/metrics       => $staging_root/metrics`
+- `pkg/controller/endpointslice/topologycache	=> $staging_root/topologycache`
```

Comment:
Tiny nit: The indentation on this line is a bit inconsistent with the rest.

Reply:
A tab snuck in somehow. Fixed.